### PR TITLE
added new function uuidv5

### PR DESCRIPTION
**Pull request by lscheidler**
it  would be nice, if this function could be added, so it can be used in different use cases.

My use-case for this function is, that it will be used in aws_route53_record resources, so that the sub-domain isn't easily guessed, but anyone knowing the dns name can convert it back to plain text.

```
resource "aws_route53_record" "client-name-or-domain" {
  zone_id = "${aws_route53_zone.primary.zone_id}"
  name    = "${uuidv5("dns", "client-name-or-domain")}.example.com"
  type    = "A"
  ttl     = "300"
  records = ["${aws_eip.lb.public_ip}"]
}
```

Rebased pull request #21197 to master branch.

### CODE REVIEW

File: lang/funcs/crypto.go [hunk 1]
```diff
@@ -32,6 +33,37 @@ var UUIDFunc = function.New(&function.Spec{
...
+		case args[0].AsString() == "dns":
+			namespace = uuidv5.NamespaceDNS
+		case args[0].AsString() == "url":
+			namespace = uuidv5.NamespaceURL
+		case args[0].AsString() == "oid":
+			namespace = uuidv5.NamespaceOID
+		case args[0].AsString() == "x500":
+			namespace = uuidv5.NamespaceX500
+		default:
+			return cty.UnknownVal(cty.String), fmt.Errorf("uuidv5() doesn't support namespace %s", args[0].AsString())
```
Comment (apparentlymart) [inline lang/funcs/crypto.go]: IIRC the namespace portion of a v5 UUID is itself a UUID, with the four you handled above as well-known ones. With that in mind, perhaps we should also allow `namespace` to be given as a literal uuid (requiring the same format that `uuid` and `uuidv5` would return) and use it verbatim if so? That way anyone who is working with a non-standard namespace they allocated themselves (or with a future standard namespace Terraform doesn't support yet) could still use it with this function.

I do like the idea of having short, readable aliases for the standard ones above, though: makes the intent much clearer to the reader in those cases than having the first argument be a literal uuid.

Comment (lscheidler) [reply lang/funcs/crypto.go]: I like this little extension and will looking into it. Thanks for Your input.

**Comment by lscheidler** [comment]
Added support for custom namespaces.

**Comment by lscheidler** [comment]
Hi @apparentlymart, any updates regarding this pull request?

**Comment by apparentlymart** [comment]
Hi @lscheidler!

Your question is timely since we're just starting to get caught up on the PRs that we had to put on hold during the v0.12 finish up.

Unfortunately I seem to have inadvertently closed this PR but I am about to merge it to `master` with just some edits to the documentation page. Sorry for the accidental close.

**Comment by apparentlymart** [comment]
Merged as aa07806bfc88506b13e0152f975d814175bfe16f. Thanks, @lscheidler!

**Pull request closed without merging**